### PR TITLE
Governance: Remove project lead role

### PR DESCRIPTION
**Pull request by yongtang**
TLDR: There was a long history of CoreDNS's governance structure (https://github.com/coredns/coredns/pull/2220)

The original reason CoreDNS had a special governance structure within CNCF (different from
most other CNCF project ), was due to the fact that Miek is the original author of
CoreDNS and is still the single biggest (more than 50%) contributor of the project.

With Miek's intention to stepping down, a possible solution
to move forward is to change the  GOVERNANCE.md similar to projects such as Envoy or Prometheus.

/cc @Tantalor93 @bradbeam @chrisohaver @darshanime @dilyevsky @ekleiner @greenpau @ihac @inigohu @isolus @jameshartig @johnbelamaric @miekg @mqasimsarfraz @nchrisdk @nitisht @pmoroney @rajansandeep @rdrozhdzh @rtreffer @snebel29 @stp-ip @superq @varyoo @ykhr53 @zouyee

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

File: GOVERNANCE.md
Comment (Tantalor93) [inline GOVERNANCE.md]:
```suggestion
If a dispute still cannot be decided, a vote can be initiated though we don't expect votes will be done very frequently.
```

**Review by Tantalor93 — COMMENTED** [review]
@yongtang so do I understand your proposal correctly, you want to remove the position of project lead?

**Comment by yongtang** [comment]
@Tantalor93 There was a long history (#2220, #2112, #2152, #2203)  of why CoreDNS set a project lead.

CoreDNS was started by Miek in 2016 single handedly. Over the years other contributors joined, some was from independent contributor community, some was part of the effort by Infoblox (as a company), and some was initially associated with Inboblox then switched to independent contributors.

(Some company may have a potential conflict of interest with CoreDNS. For example, Infoblox does sell other DNS server/product that may compete with CoreDNS)

However, Miek has always been the biggest contributor of the project  since the beginning, as you can see from the highlight.

Miek himself is serving CoreDNS community as an independent contributor. (See [Original OWNDERS file](https://github.com/coredns/coredns/blob/669b99d91c5eaf4c669d03436083c98e3c298367/OWNERS) )

When CoreDNS intended to join CNCF and set the Governance, the initial idea was to form a structure that more or less aligned with other project like Prometheus or Envoy. However, there was concerns (mostly from independent contributor community) that company might dominate the CoreDNS project (commercially) and independent contributors may be pushed out (See #2220, #2112, #2152, #2203).

At one point it was proposed that Miek will be served as BDFL (Benevolent-dictator-for-life) of CoreDNS (See https://github.com/coredns/coredns/pull/2152).

This was a reasonable solution, because Miek is effectively and undeniably the BDFL for CoreDNS project. And he is serving as an independent contributor from community (avoid conflict of interest).

Some concern was raised by Francois from Infoblox as "for-life" seems to be a strong term. I think this is a valid concern.

Eventually the compromise (#2220) was that a "project lead" will be used with a term of 1 year (voted by maintainers every year). But to be honest, we understand Miek is the only person suitable for the "project lead" role (Benevolent-dictator-not-for-life). 

With Miek stepping down, I do believe removing the  "project lead" role (Benevolent-dictator-not-for-life) will help CoreDNS moving forward:

This structure avoids potential conflict in between different organizations, companies, and independent contributors, and align with other projects in CNCF such as Envoy or Prometheus.

This is also a respect to Miek's unique contribution for CoreDNS project - I believe he should be the project lead whenever he want.

**Comment by Tantalor93** [comment]
@yongtang thanks for detailed clarification👍It makes sense to move CoreDNS governance this way as CoreDNS is now quite mature project, but my vote is quite minor as I am quite fresh maintainer 🙂

**Comment by greenpau** [comment]
> This is also a respect to Miek's unique contribution for CoreDNS project - I believe he should be the project lead whenever he want.

💯 👍

**Review by dilyevsky — APPROVED** [review]

**Review by bradbeam — APPROVED** [review]

**Review by greenpau — APPROVED** [review]

**Review by snebel29 — APPROVED** [review]
Sounds good to me.

**Review by zouyee — APPROVED** [review]

**Review by nitisht — APPROVED** [review]

**Review by darshanime — APPROVED** [review]

**Review by inigohu — APPROVED** [review]

**Review by rtreffer — APPROVED** [review]

**Review by ihac — APPROVED** [review]

**Comment by johnbelamaric** [comment]
I agree it makes sense to move away from a single leader. I do not, however, think a completely flat structure will work well for us. The primary reason being that we have a very low bar to become a maintainer. There are some things - for example, admin privileges on the GitHub org - that I do not believe should be made available to all maintainers due to that low bar. Not that there is anyone I am concerned about, but I think it's important to minimize that sort of access.

What I suggest instead would be a 3 (or maybe 5?) person steering committee that replaces the currently leader role. We would elect those members in the same manner we currently elect the leader, with the same cadence. That committee could then effectively make the same sort of decisions the leader would have made, except by vote. We can also limit things like org admin access to those folks or their delegates.

WDYT?

**Comment by chrisohaver** [comment]
> I do not, however, think a completely flat structure will work well for us. 

Indeed, even the the example projects cited in the description (Envoy and Prometheus), do not in fact implement a flat structure.  Both implement a multi-tiered system, where a subset or subsets of maintainers have elevated roles and or distinct responsibilities (specifically as it relates to decision making).

**Comment by chrisohaver** [comment]
I have updated the title of this PR to be more specific to what the changes are proposing.  I hope this is OK!

**Comment by SuperQ** [comment]
Do we want to go with this or the steering committee one?

Funny enough you mention Prometheus. We're currently considering moving to a steering committee style ourselves so that we can onboard more maintainers, but reduce the effect of lots of maintainers from one organization.

**Comment by yongtang** [comment]
> Do we want to go with this or the steering committee one?

@SuperQ  Let's leave both and let other maintainers weigh in what is the best way (committee vs. flat) for the community.

In case a steering committee is preferred (vs. a simple vote from all maintainers), we may also need to see what the structure will be. For example, from what I can see, Envoy has a ["senior maintainer" concept](https://github.com/envoyproxy/envoy/blob/main/GOVERNANCE.md#conflict-resolution-and-voting) where each  "senior maintainer" have 2 votes while others have 1 vote. So this might also be something to consider.

**Comment by chrisohaver** [comment]
> For example, from what I can see, Envoy has a ["senior maintainer" concept](https://github.com/envoyproxy/envoy/blob/main/GOVERNANCE.md#conflict-resolution-and-voting) where each "senior maintainer" have 2 votes while others have 1 vote. So this might also be something to consider.

This seems more compatible with a no-leader approach rather than a steering committee approach.  However IIUC, Envoy also expects that all of their maintainers dedicate 25% of their time to the project.  This is IMO a very high bar, but the result is that all votes (from senior or otherwise) are from highly engaged maintainers.

**Comment by yongtang** [comment]
If I remember, CNCF has 16 graduated projects with each projects have different ways of organize. We don't necessarily follow any specific project out of all CNCF projects though we can certainly learn from different projects and see what is the best way for CoreDNS community.

**Comment by chrisohaver** [comment]
I think it's fair to say that in essence, this change is replacing final decisions by leader with final decision by maintainer vote.

IMO, it's best for the CoreDNS community if all final decision makers are actively engaged, have good present knowledge of CoreDNS as a whole, and ideally have expert knowledge of DNS. But I don't think that we can expect _all_ maintainers to have those qualities.  

Per the governance, the only way a maintainer can lose that status is if they voluntarily leave, or violate Code of Conduct. On its own, it is not a problem at all for a maintainer to be perpetually uninvolved. But it does mean that over time, we can accumulate a long list of unengaged maintainers, and I think anyone who is involved in final decision votes should be actively engaged in the project.  

As far as maintainer expectations go, aside from code of conduct we currently have one expectation defined in the governance: that all maintainers should respond to PRs within a "reasonable" time.  We won't eject a maintainer for not meeting maintainer expectations, _and I'm not suggesting we should_.  But IMO, a final decision maker should at least meet this expectation.  But "reasonable time" is subjective, and hard to determine on a  per maintainer basis. Therefore it's not feasible to ensure that final decision voting is only done by maintainers who meet maintainer expectations.

**Review by johnbelamaric — CHANGES_REQUESTED** [review]
-1
Prefer the steering committee as described in #5531

**Comment by chrisohaver** [comment]
-1

**Comment by inigohu** [comment]
-1

**Comment by johnbelamaric** [comment]
It has come to our attention that not everyone is on the maintainers@coredns.io mailing list over which we have been communicating about the two governance PRs. Please vote -1/+1 on each of these PRs: #5494 and #5531.

@bradbeam
@chrisohaver
@darshanime
@dilyevsky
@ekleiner
@greenpau
@ihac
@inigohu
@isolus
@jameshartig
@johnbelamaric
@miekg
@mqasimsarfraz
@nchrisdk
@nitisht
@pmoroney
@rajansandeep
@rdrozhdzh
@rtreffer
@snebel29
@stp-ip
@superq
@Tantalor93
@varyoo
@ykhr53
@yongtang
@zouyee

**Comment by johnbelamaric** [comment]
Please vote by tomorrow (Thursday, September 8th) noon pacific time, if you have not already done so.

**Review by rajansandeep — COMMENTED** [review]
-1

**Comment by snebel29** [comment]
Having given a second thought I would like to change my vote giving -1 to this proposal and +1 https://github.com/coredns/coredns/pull/5531 instead.

**Comment by mqasimsarfraz** [comment]
-1

**Review by Isolus — COMMENTED** [review]
-1

**Comment by SuperQ** [comment]
-1

**Comment by johnbelamaric** [comment]
Thank you everyone, we have a winner.

Final voting results:
 - No Lead: 6
 - Steering: 14
 - Abstention: 6


Maintainer | No Lead (#5494) | Steering (#5531) | Choice
-- | -- | -- | --
@bradbeam | +1 (2022-07-07) | +1 (2022-09-07) | Steering
@chrisohaver | -1 (2022-08-31) | +1 (2022-08-31) | Steering
@darshanime | +1 (2022-07-09) |   | No Lead
@dilyevsky | +1 (2022-07-07) | +1 (2022-07-26) | Steering
@ekleiner |   |   |  
@greenpau | +1 (2022-07-07) |   | No Lead
@ihac | +1 (2022-07-12) | +1 (2022-09-07) | Steering
@inigohu | -1 (2022-08-31) | +1 (2022-08-31) | Steering
@isolus | -1 (2022-09-07) | +1 (2022-09-07) | Steering
@jameshartig |   |   |  
@johnbelamaric | -1 (2022-08-09) | +1 (2022-08-09) | Steering
@miekg |   |   |  
@mqasimsarfraz | -1 (2022-09-07) | +1 (2022-09-07) | Steering
@nchrisdk |   |   |  
@nitisht | +1 (2022-07-08) |   | No Lead
@pmoroney |   | +1 (2022-08-31) | Steering
@rajansandeep | -1 (2022-09-07) | +1 (2022-09-01) | Steering
@rdrozhdzh |   |   |  
@rtreffer | +1 (2022-07-09) |   | No Lead
@snebel29 | -1 (2022-09-07) | +1 (2022-09-07) | Steering
@stp-ip |   | +1 (2022-09-01) | Steering
@superq | -1 (2022-09-07) | +1 (2022-09-07) | Steering
@Tantalor93 |   | +1 (2022-07-28) | Steering
@varyoo |   |   |  
@ykhr53 |   |   |  
@yongtang | +1 (2022-07-07) |   | No Lead
@zouyee | +1 (2022-07-08) |   | No Lead

**Pull request closed without merging**
